lecture-notes/3-bld.md

```diff
@@ -323,3 +328,156 @@ Make has been around a long time and its performance and correctness are
 are at home in most other projects that use Make. Not so for a custom shell
 script.
+
+### Variables
+
+Up until now we have written rules and filenames and commands as constants in
+the Makefile. This means that if you ever had to reference a file `foo.c` more
+than once in the Makefile, you would have to copy around the text. Fortunately,
+like other programming languages, Make provides variables.
+
+User-defined variables come in two forms in Make: *simply expanded* variables
+use `NAME := val` syntax and *recursively expanded* variables use `NAME = val`
+syntax. You will likely not want to use recursively expanded variables much, if
+at all, as they have different semantics than in other programming languages:
+
+**Simply expanded** means that variable references and function calls on the
+right hand side happen when the variable is *defined* (in the *read-in* phase).
+
+In the below example, we have a simply expanded variable `EXE` that can be used
+throughout the Makefile to refer to the desired executable name. This allows
+for a single point of truth.
+
+```make
+EXE := main
+$(EXE): main.o log.o
+	gcc main.o log.o -o $(EXE)
+main.o: main.c log.h
+	gcc -c main.c
+log.o: log.c log.h
+	gcc -c log.c
+.PHONY: clean
+clean:
+	rm -rf *.o $(EXE)
+```
+
+(While variables can normally have any name, the convention is to use
+upper-case characters.)
+
+**Recursively expanded** means that variable references and function calls on
+the right hand side happen when the variable is *used* (in the *target-update*
+phase for recipe usage). This structure allows complex templates to be stored
+in a variable and used with many different values.
+
+Expansions of both kinds of variables can occur anywhere, and use `$(VAR)`
+syntax.
+
+In the below example, we have a mix of simply and recursively expanded
+variables. `LOG_TARGET` is recursively expanded (note the `=`) and `CFLAGS` is
+simply expanded (note the `:=`).
+
+```make
+LOG_TARGET = echo "Current target: $@" ; echo "Prerequisites: $^"
+CFLAGS := -Wall
+CFLAGS := $(CFLAGS) -Werror
+main: main.o log.o
+	$(LOG_TARGET)
+	gcc $(CFLAGS) $^ -o $@
+main.o: main.c log.h
+	$(LOG_TARGET)
+	gcc $(CFLAGS) -c $< -o $@
+log.o: log.c log.h
+	$(LOG_TARGET)
+	gcc $(CFLAGS) -c $< -o $@
+```
+
+Because `LOG_TARGET` is recursively expanded, it is expanded at each use site
+in the Makefile: in `main`, in `main.o`, and in `log.o`. Every time it is
+expanded, the variable references inside it see the automatic variables (those
+funky-looking variables) local to that rule.
+
+Speaking of automatic variables...
+
+### Automatic variables
+
+Similar to how C gives you `__FILE__` and `__LINE__` and `__func__`, Make also
+defines some variables for you within rules. These special variables are called
+*automatic variables* and help refer to different parts of the rule so you, the
+programmer, don't repeat yourself too much.
+
+There are many automatic variables, particuarly in the GNU implementation of
```

Comment:
Should we link to the GNU docs that have the full list here? Can't remember if that's our practice in lecture notes or not.